harness_tests/diagnostic_test/main1.c

```diff
@@ -1,205 +1,478 @@
 /**
- * Diagnostic test to verify hardware functionality for EPS. This consists of: 
- *      - taking measurements with devices and making sure they are within 
- *          expected range
- *      - turning on a heater shuld increase current from boost and battery
- *      - imu gyro vals should be small numbers
- */
+ * Diagnostic test to verify hardware functionality for EPS.
+ *      - See Bus Architecture Specification (section 5- diagnostic tests) for more information
+ **/
+
 #include <stdbool.h>
 
 #include <adc/adc.h>
 #include <conversions/conversions.h>
 #include <uart/uart.h>
 #include <test/test.h>
+#include <can/data_protocol.h>
 
 #include "../../src/devices.h"
 #include "../../src/heaters.h"
 #include "../../src/imu.h"
 #include "../../src/measurements.h"
 #include "../../src/shunts.h"
-
-
-double read_voltage(uint8_t channel) {
-    fetch_adc_channel(&adc, channel);
-    uint16_t raw_data = read_adc_channel(&adc, channel);
-    double voltage = adc_raw_to_circ_vol(raw_data, ADC_VOL_SENSE_LOW_RES, ADC_VOL_SENSE_HIGH_RES);
-    return voltage;
+#include "../../src/can_commands.h"
+
+uint8_t tx_msg[8] = {0x00};
+
+/* Helper function for generating and appropriately dequeueing rx_housekeeping message */
+void construct_rx_msg_hk(uint8_t field_num){
+    uint8_t rx_msg[8] = {0x00};
+    rx_msg[2] = CAN_EPS_HK;
+    rx_msg[3] = field_num;
+    enqueue(&can_rx_msg_queue, rx_msg);
+    handle_rx_msg();
+    dequeue(&can_tx_msg_queue, tx_msg);
 }
 
-double read_current(uint8_t channel) {
-    fetch_adc_channel(&adc, channel);
-    uint16_t raw_data = read_adc_channel(&adc, channel);
-    double current = adc_raw_to_circ_cur(raw_data, ADC_DEF_CUR_SENSE_RES, ADC_DEF_CUR_SENSE_VREF);
-    return current;
+/* Helper function for generating and appropriately dequeueing rx_ctrl message */
+void construct_rx_msg_ctrl(uint8_t field_num, uint32_t data){
+    uint8_t rx_msg[8] = {0x00};
+    rx_msg[2] = CAN_EPS_CTRL;
+    rx_msg[3] = field_num;
+    tx_msg[4] = (data >> 24) & 0xFF;
+    tx_msg[5] = (data >> 16) & 0xFF;
+    tx_msg[6] = (data >> 8) & 0xFF;
+    tx_msg[7] = data & 0xFF;
+    enqueue(&can_rx_msg_queue, rx_msg);
+    handle_rx_msg();
+    dequeue(&can_tx_msg_queue, tx_msg);
 }
 
-double read_bat_current(uint8_t channel) {
-    fetch_adc_channel(&adc, channel);
-    uint16_t raw_data = read_adc_channel(&adc, channel);
-    double current = adc_raw_to_circ_cur(raw_data, ADC_BAT_CUR_SENSE_RES, ADC_BAT_CUR_SENSE_VREF);
+/* Helper function for measuring heater currents */
+double measure_heater_current(uint8_t source){
+    construct_rx_msg_hk(source);
+    uint16_t raw_data = (tx_msg[4] << 8) & tx_msg[5];
```

Comment:
You would use a bitwise OR (`|`) instead of AND (`&`) to get the data properly. Also it should be 4 bytes and just get it as the return value from `construct_rx_msg_hk()` (see my comments above)